services/cargohold/src/CargoHold/API/Public.hs

```diff
@@ -132,13 +144,36 @@ mkAssetLocation key =
 uploadAssetV3 ::
   forall tag id.
   MakePrincipal tag id =>
+  Endpoint ->
   id ->
   AssetSource ->
   Handler (Asset, AssetLocation Relative)
-uploadAssetV3 pid req = do
+uploadAssetV3 (Endpoint h p) pid req = do
   let principal = mkPrincipal pid
+      baseUrl = BaseUrl Http (T.unpack h) (fromIntegral p) ""
+  clientEnv <- liftIO $ newManager defaultManagerSettings <&> \mgr -> ClientEnv mgr baseUrl Nothing defaultMakeClientRequest
+  let getUser uid = runClientM (client' uid) clientEnv
```

Comment:
We should do all of this at boot time for cargohold and keep it in the environment. This way we can avoid having to pass this `Endpoint` here and having to thread it through all the functions above.

Reply:
Done, I've moved the creation of the ClientEnv into the environment for the Handler.